arch/common/init.c.tpl

```diff
@@ -62,10 +57,32 @@ static void *load_library() {
 #endif
 
   do_dlclose = 1;
+
+  return lib_handle;
+}
+
+#if __cplusplus > 201103L
```

Comment:
I think we need `__cpp_threadsafe_static_init` (to handle GCC's pre-C++11 `-fthreadsafe-statics`) and `! LAZY_LOAD` here.

Reply:
I'm not sure I understand why we would not need to guard library loading when `LAZY_LOAD` is set. 
While chances of a race are probably lower, the race is still possible.

Reply:
Sorry, I didn't realize that `do_load_library` will be called at first call to `load_library`, not at program startup.

The solution is great except that it will only work for C++ code. Do you think `pthread_once` could be used instead, to cover C and pre-C++11?

Reply:
I personally would just switch to `C++`. Are there particular reasons that init code has to be plain `C`?

Reply:
I think we shouldn't impose C++ dependency on users with C codebases.

Reply:
By the same logic, we should not impose `pthreads` on those who don't or can't use it. It's a judgement call either way. Considering that it's your code, it's your call in the end. 
I can add pthreads-based init when I'm back to work. Or it can be done as a separate improvement -- my patch and making library load thread safe for C users are independent and do not have to be done in one patch.

Reply:
> By the same logic, we should not impose pthreads on those who don't or can't use it.

True but pthread integration is simpler (just adding `-pthread` to `LDFLAGS`) whereas adding C++ code would require users to support two languages in their Makefiles and add C++ compiler dependency in distro package scripts.